dataplane/ebpf/src/ingress/tcp.rs

```diff
@@ -116,21 +126,106 @@ pub fn handle_tcp_ingress(ctx: TcContext) -> Result<i32, i64> {
         return Ok(TC_ACT_OK);
     }
 
-    // Calculate l3 cksum
-    // TODO(astoycos) use l3_cksum_replace instead
-    unsafe { (*ip_hdr).check = 0 };
-    let full_cksum = unsafe {
-        bpf_csum_diff(
-            mem::MaybeUninit::zeroed().assume_init(),
+    let mut lb_mapping = LoadBalancerMapping {
+        backend,
+        backend_key,
+        tcp_state,
+    };
+
+    let tcp_hdr_ref = unsafe { tcp_hdr.as_ref().ok_or(TC_ACT_OK)? };
+
+    // If the packet has the RST flag set, it means the connection is being terminated, so remove it
+    // from our map.
+    if tcp_hdr_ref.rst() == 1 {
+        unsafe {
+            LB_CONNECTIONS.remove(&client_key)?;
+        }
+    }
+
+    update_tcp_conns(tcp_hdr_ref, &client_key, &mut lb_mapping)?;
```

Comment:
Should this be moved to after the cksum calculation code block as it was prior to this patch?

Reply:
in fact, I found that I cannot put

```rust
    if tcp_hdr_ref.rst() == 1 {
        unsafe {
            LB_CONNECTIONS.remove(&client_key)?;
        }
    }

    let mut lb_mapping = LoadBalancerMapping {
        backend,
        backend_key,
        tcp_state,
    };

    update_tcp_conns(tcp_hdr_ref, &client_key, &mut lb_mapping)?;
```

behind 

```rust
    let backend_ip = backend.daddr.to_be();
    let ret = set_ipv4_ip_dst(&ctx, TCP_CSUM_OFF, &original_daddr, backend_ip);
    if ret != 0 {
        return Ok(TC_ACT_OK);
    }

    let backend_port = (backend.dport as u16).to_be();
    let ret = set_ipv4_dest_port(&ctx, TCP_CSUM_OFF, &original_dport, backend_port);
    if ret != 0 {
        return Ok(TC_ACT_OK);
    }
```

otherwise eBPF program would load failed, eBPF verifier shows that access out of bounds

Reply:
I think accessing `tcp_hdr` after `bpf_redirect_neigh` will cause an out-of-bounds error, but I don't know why the original code did not have this issue.